src/sympc/tensor/replicatedshare_tensor.py

```diff
@@ -209,6 +249,138 @@ def ne(self, y):
+        share_ptrs: List["ReplicatedSharedTensor"],
+        get_shares: bool = False,
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        """Reconstruct value from shares.
+
+        Args:
+            share_ptrs (List[ReplicatedSharedTensor]): List of RSTensor pointers.
+            get_shares (bool): Retrieve only shares.
+
+        Returns:
+            reconstructed_value (torch.Tensor): Reconstructed value.
+
+        Raises:
+            ValueError: When parties share values are not equal.
+        """
+        nparties = len(share_ptrs)
+
+        # Get shares from all parties
+        all_shares = []
+        for party_rank in range(nparties):
+            all_shares.append(share_ptrs[party_rank].get_shares().get())
+
+        # reconstruct shares from all parties and verify
+        value = None
+        for party_rank in range(nparties):
+            share_sum = all_shares[party_rank][0] + sum(
+                all_shares[(party_rank + 1) % (nparties)]
+            )
+
+            if not value:
+                value = share_sum
+
+            elif share_sum != value:
+                raise ValueError(
+                    "Reconstruction values from all parties are not equal."
+                )
+
+        if get_shares:
+            return all_shares
+
+        return value
+
+    @staticmethod
+    def reconstruct(
+        share_ptrs: List["ReplicatedSharedTensor"],
+        get_shares: bool = False,
+        security_type: str = "semi-honest",
+    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        """Reconstruct value from shares.
+
+        Args:
+            share_ptrs (List[ReplicatedSharedTensor]): List of RSTensor pointers.
+            security_type (str): Type of security followed by protocol.
+            get_shares (bool): Retrieve only shares.
+
+        Returns:
+            reconstructed_value (torch.Tensor): Reconstructed value.
+
+        Raises:
+            ValueError: Invalid security type
+        """
+        if security_type == "malicious":
+            return ReplicatedSharedTensor.__reconstruct_malicious(
+                share_ptrs, get_shares
+            )
+
+        elif security_type == "semi-honest":
+            return ReplicatedSharedTensor.__reconstruct_semi_honest(
+                share_ptrs, get_shares
+            )
+
+        raise ValueError("Invalid security Type")
+
+    @staticmethod
+    def distribute_shares(shares: List[ShareTensor], session: Session) -> List:
+        """Distribute a list of shares.
+
+        Args:
+            shares (List[ShareTensor): list of shares to distribute.
+            session (Session): Session.
+
+        Returns:
+            List of ShareTensor.
+
+        """
+        parties = session.parties
+
+        nshares = len(parties) - 1
+
+        ptr_list = []
+        for i in range(len(parties)):
+            party_ptrs = []
+
+            for j in range(i, i + nshares):
+                tensor = shares[j % (nshares + 1)].tensor
```

Comment:
I think we should use here `nparties` instead `nshares`
Similar like : `tensor = shares[j % nparties].tensor` like you did for all shares
https://github.com/OpenMined/SyMPC/blob/fe974c98eaf506060f3148ed6d213c2e7be69488/src/sympc/tensor/replicatedshare_tensor.py#L304
We made an explicit assumption that 
nshares= nparties-1
I think for generalized RSS ,its bit different.

Reply:
No, it's the same for generalized RSTensor.

Reply:
I believe that it is protocol dependent Hrishi.Considering a scenario of 5 parties.If the share distribution is done according to our setting.`nshares`= `nparties-1`. The protocol could not guarantee malicious security for honest majority as two parties can always collude to reveal the private input.I think such protocol would do the share distribution differently to ensure honest majority. I might be wrong here too.

Reply:
Okay, let's change it later I guess. Right now for Falcon and ABY3 it works just fine.

Reply:
Clearly defining what this behaviour will be like will require a good amount of literature survey and we aren't sure if we will use all of them. As long as this is good enough for a default protocol, we are good.